File: FactoryMethod/factory.ts
interface Transport{
    delivery(): string;
}

export class Car implements Transport{
    delivery(): string {
        return "Delivery by car";
    }
}

export class Truck implements Transport{
     delivery(): string {
        return "Delivery by truck";
    }
}


abstract class TransportFactory {

    abstract factoryMethod(): Transport;

    someOperation(): string {
        const transport =  this.factoryMethod();
        return transport.delivery();
    }

}

class CarFactory extends  TransportFactory {

    factoryMethod(): Transport {
        return new Car();
    }

}


class TruckFactory extends TransportFactory{

    factoryMethod(): Transport {
        return new Truck();
    }

}

const someCode = (factory:  TransportFactory) =>{
    console.log(factory.someOperation());
}


someCode(new CarFactory());
someCode(new TruckFactory());

File: CreationalPatterns/FactoryMethod/factory.cs
using System;


    public interface ITransport
    {
        string Delivery();
    }

    public class Car : ITransport
    {
        public string Delivery()
        {
            return "Delivery by Car";
        }
    }

    public class Truck : ITransport
    {
        public string Delivery()
        {
            return "Delivery by Truck";
        }
    }

    
    public abstract class TransportFactory
    {
        public abstract ITransport FactoryMethod();

        public string SomeOperation()
        {
            var transport = this.FactoryMethod();
            return transport.Delivery();
        }
    }

    public class CarFactory : TransportFactory
    {
        public override ITransport FactoryMethod()
        {
            return new Car();                        
        }
    }

    public class TruckFactory : TransportFactory
    {
        public override ITransport FactoryMethod()
        {
            return new Truck();
        }
    }

    public class Program
    {
        static void Main(string[] args)
        {
                
            TransportFactory factory = new CarFactory();
            TransportFactory transport = new TruckFactory();

            Console.WriteLine(factory.SomeOperation());
            Console.WriteLine(transport.SomeOperation());
            Console.WriteLine(factory.FactoryMethod());
            Console.WriteLine(transport.FactoryMethod());
                
                

        }
        
        
    }


